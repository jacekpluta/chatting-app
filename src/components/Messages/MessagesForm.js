import { Segment, Input, Button } from "semantic-ui-react";
import React, { useState, useEffect } from "react";
import firebase from "../Firebase";
import mime from "mime-types";
import ModalFile from "./ModalFile";
import { uuid } from "uuidv4";

export default function MessagesForm(props) {
  const [storageRef] = useState(firebase.storage().ref());

  const [message, setMessage] = useState();
  const [loading, setLoading] = useState(false);

  const [error, setError] = useState("");
  const [modal, setModal] = useState(false);
  const [uploadState, setUploadState] = useState("");
  const [uploadTask, setUploadTask] = useState(null);
  const [file, setFile] = useState(null);
  const [authorized] = useState(["image/jpeg", "image/png"]);

  const {
    messagesRef,
    currentChannel,
    currentUser,
    isPrivateChannel,
    getMessagesRef
  } = props;

  const handleChange = event => {
    setMessage(event.target.value);
  };

  const createMessage = (downloadURL = null) => {
    const createMessage = {
      timeStamp: firebase.database.ServerValue.TIMESTAMP,
      currentUser: {
        id: currentUser.uid,
        name: currentUser.displayName,
        avatar: currentUser.photoURL
      }
    };
    if (downloadURL !== null) {
      createMessage.image = downloadURL;
    } else {
      createMessage.content = message;
    }
    return createMessage;
  };

  const openModal = () => {
    setModal(true);
  };

  const closeModal = () => {
    setModal(false);
  };

  const sendMessage = () => {
    if (message) {
      setLoading(true);
      getMessagesRef()
        .child(currentChannel.id)
        .push()
        .set(createMessage())
        .then(() => {
          setLoading(false);
          setMessage("");
          setError("");
        })
        .catch(error => {
          console.log(error);
          setLoading(false);
          setMessage("");
          setError(error);
        });
    } else {
      setError("Add a message");
    }
  };

  const handleChangeFileUpload = event => {
    const file = event.target.files[0];

    if (file) {
      setFile(file);
    }
  };

  const sendFile = () => {
    if (file && isAuthorized()) {
      const metadata = { conetentType: mime.lookup(file.name) };
      uploadFile(file, metadata);
    } else {
      closeModal();
    }
  };

  const isAuthorized = () => {
    closeModal();
    setFile(null);
    return authorized.includes(mime.lookup(file.name));
  };

  const getPath = () => {
    if (currentUser) {
      return `chat/private-${currentChannel.id}`;
    } else {
      return `chat/public`;
    }
  };

  const uploadFile = (file, metadata) => {
    console.log(getPath());
    const pathToUpload = currentChannel.id;
    const messRef = getMessagesRef();
    const filePath = `${getPath()}/${uuid()}.jpg`;

    setUploadState("uploading");

    props.setMessageImageLoadingTrue();

    const task = storageRef
      .child(filePath)
      .put(file, metadata)
      .then(snapshot => {
        setUploadTask(task);
        snapshot.ref.getDownloadURL().then(downloadURL => {
          sendFileMessage(downloadURL, messRef, pathToUpload);
        });
      })
      .then(() => {
        props.setMessageImageLoadingFalse();
      })
      .catch(error => {
        setUploadState("error");
        setUploadTask(null);
        console.log(error);
      });
  };

  useEffect(() => {
    if (uploadTask) {
      uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(
        snapshot
      ) {
        var percent = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
        console.log(percent + "% done");
      });
    }
  }, []);

  const sendFileMessage = (downloadURL, messRef, pathToUpload) => {
    messRef
      .child(pathToUpload)
      .push()
      .set(createMessage(downloadURL))
      .then(() => {
        setUploadState("done");
      })
      .catch(error => {
        setUploadState("error");
        setUploadTask(null);
        setError(error);
        console.log(error);
      });
  };

  return (
    <React.Fragment>
      <Segment className="messageForm">
        <Input
          value={message}
          fluid
          name="message"
          style={{ marginBottom: "0.7 em" }}
          label={<Button icon={"add"}></Button>}
          placeholder="Write your message"
          onChange={handleChange}
          className={error.includes("message") ? "error" : ""}
        />
        <Button.Group icon widths="2">
          <Button
            disabled={loading}
            onClick={sendMessage}
            color="orange"
            content="Add Reply"
            labelPosition="left"
            icon="edit"
          ></Button>
          <Button
            disabled={props.messageImageLoading}
            onClick={openModal}
            color="teal"
            content="Upload Media"
            labelPosition="right"
            icon="upload"
          ></Button>
        </Button.Group>
      </Segment>
      <ModalFile
        modal={modal}
        closeModal={closeModal}
        error={error}
        handleChangeFileUpload={handleChangeFileUpload}
        sendFile={sendFile}
      ></ModalFile>
    </React.Fragment>
  );
}
