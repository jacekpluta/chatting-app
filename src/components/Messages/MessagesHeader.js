import React, { useState } from "react";

import {
  Header,
  Segment,
  Input,
  Icon,
  Popup,
  List,
  Image,
} from "semantic-ui-react";
import MetaPanel from "../SidePanel/MetaPanel/MetaPanel";
import DeleteChannelModal from "../SidePanel/MetaPanel/DeleteChannelModal";
import { connect } from "react-redux";
import firebase from "../Firebase";
import {
  setCurrentChannel,
  setActiveChannelId,
  setPrivateChannel,
} from "../../actions";

const MessagesHeader = (props) => {
  const {
    displayChannelName,
    handleSearchChange,
    searchLoading,
    isPrivateChannel,
    handleStarred,
    channelStarred,
    unstarChannel,
    handleAddFriend,
    handleUnfriendPerson,
    currentChannel,
    userPosts,
    currentUser,
    usersInChannel,
    channelFriended,
  } = props;

  const [popupOpen, setPopupOpen] = useState(false);
  const [loading, setLoading] = useState(false);
  const [channelsRef] = useState(firebase.database().ref("channels"));
  const [openModal, setOpenModal] = useState(false);

  const handleDisablePopup = () => {
    setPopupOpen(false);
  };

  const handleEnablePopup = () => {
    setPopupOpen(true);
  };

  const handleCloseModal = () => {
    setOpenModal(false);
  };

  const handleOpenModal = () => {
    setOpenModal(true);
    handleDisablePopup();
  };

  const handleDeleteChannel = () => {
    setLoading(true);

    const mainChannel = {
      id: "mainChannel",
      name: "main channel",
      details: "This is main channel",
      createdBy: {
        uid: "111",
        name: "Admin",
        avatar:
          "https://firebasestorage.googleapis.com/v0/b/react-chatting-app-4d9cb.appspot.com/o/avatars%2Fadmin.jpg?alt=media&token=bbcb9b2d-d338-4f02-9fa3-0b2b309c0695",
      },
    };

    channelsRef
      .child(currentChannel.id)
      .remove((err) => {
        if (err !== null) {
          console.log(err);
        }
      })
      .then(() => {
        setLoading(false);
        props.setCurrentChannel(mainChannel);
        props.setActiveChannelId(mainChannel.id);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const getChannelId = (userId) => {
    const currentUserId = currentUser.uid;
    return userId < currentUserId
      ? `${userId}/${currentUser.uid}`
      : `${currentUser.uid}/${userId}`;
  };

  const changeChannel = (user) => {
    const channelId = getChannelId(user.id);

    const privateChannelData = {
      id: channelId,
      name: user.name,
      photoURL: user.avatar,
    };

    props.setActiveChannelId(user.id);
    props.setCurrentChannel(privateChannelData);
    props.setPrivateChannel(true);
  };

  //CHECKS IF THERE ARE USERS IN CURRENT CHANNEL
  const isEmpty = (obj) => {
    for (let key in obj) {
      if (obj.hasOwnProperty(key)) return false;
    }
    return true;
  };

  //DISPLAYS CURRENT USERS IN CHANNEL
  const displayUsersInChannel = () =>
    usersInChannel
      .sort((a, b) => (a.name > b.name ? 1 : -1))
      .map((user) => (
        <List>
          <List.Item
            key={user.uid}
            name={user.name}
            onClick={() => changeChannel(user)}
            // active={!favouriteActive && activeChannelId === channel.id}
          >
            <Image avatar src={user.avatar} />
            <List.Content>
              <List.Header as="a">
                {" "}
                <span> # {user.name}</span>
              </List.Header>
              <List.Description>In channel</List.Description>
            </List.Content>
          </List.Item>
        </List>
      ));

  return (
    <Segment clearing className="messagesHeader">
      <Header
        style={{ paddingTop: "16px" }}
        fluid="true"
        as="h3"
        floated="left"
      >
        <span style={{ fontWeight: "bold" }}>
          {" "}
          {displayChannelName()}{" "}
          {!isPrivateChannel && channelStarred && (
            <Icon
              name={"star"}
              onClick={unstarChannel}
              className="iconColor"
            ></Icon>
          )}
          {!isPrivateChannel && !channelStarred && (
            <Icon
              name={"star outline"}
              onClick={handleStarred}
              className="iconColor"
            ></Icon>
          )}
          {isPrivateChannel && channelFriended && (
            <Icon
              name={"user times"}
              onClick={handleUnfriendPerson}
              color="red"
            ></Icon>
          )}
          {isPrivateChannel && !channelFriended && (
            <Icon
              name={"user plus"}
              onClick={handleAddFriend}
              color="green"
            ></Icon>
          )}
        </span>
      </Header>
      <Header
        as="h3"
        floated="right"
        textAlign="center"
        style={{ paddingTop: "6px" }}
        fluid="true"
      >
        <span>
          {!isPrivateChannel && (
            <Popup
              onOpen={handleEnablePopup}
              onClose={handleDisablePopup}
              open={popupOpen}
              position="bottom center"
              flowing
              hoverable
              trigger={<Icon className="iconColor" name="question circle" />}
            >
              <MetaPanel
                key={currentChannel && currentChannel.id}
                currentChannel={currentChannel}
                isPrivateChannel={isPrivateChannel}
                userPosts={userPosts}
                currentUser={currentUser}
                handleDisablePopup={handleDisablePopup}
                handleOpenModal={handleOpenModal}
                loading={loading}
              ></MetaPanel>
            </Popup>
          )}
          {"  "}
          {!isPrivateChannel && currentChannel && (
            <Popup
              position="bottom center"
              style={{ paddingRight: "10px" }}
              flowing
              hoverable
              trigger={<Icon className="iconColor" name="users" />}
            >
              {usersInChannel ? displayUsersInChannel() : ""}
              {isEmpty(usersInChannel) ? (
                <List>
                  <List.Item>
                    <List.Content>
                      <List.Header as="a">
                        {" "}
                        <span style={{ color: "#00000" }}>
                          No users in channel
                        </span>
                      </List.Header>
                    </List.Content>
                  </List.Item>
                </List>
              ) : (
                ""
              )}
            </Popup>
          )}
          {"    "}
          <Input
            onChange={handleSearchChange}
            icon="search"
            name="searchTerm"
            loading={searchLoading}
            placeholder={`search ${currentChannel ? currentChannel.name : ""}`}
            style={{ width: "170px" }}
          ></Input>
        </span>
      </Header>

      <DeleteChannelModal
        handleDeleteChannel={handleDeleteChannel}
        handleCloseModal={handleCloseModal}
        openModal={openModal}
        loading={loading}
      />
    </Segment>
  );
};

export default connect(null, {
  setCurrentChannel,
  setActiveChannelId,
  setPrivateChannel,
})(MessagesHeader);
