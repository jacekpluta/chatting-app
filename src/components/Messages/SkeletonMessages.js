import React from "react";
import Skeleton from "react-loading-skeleton";
import { Grid } from "semantic-ui-react";

export default function SkeletonMessages() {
  return Array(10)
    .fill()
    .map((item, index) => (
      <Grid columns={2}>
        <Grid.Row>
          <Grid.Column width={1}>
            <Skeleton square={true} height={50} width={50} />
          </Grid.Column>

          <Grid.Column width={3}>
            <Grid.Row>
              <Skeleton />
            </Grid.Row>
            <Grid.Row style={{ paddingTop: 5 }}>
              <Skeleton height={20} width={580} />
            </Grid.Row>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    ));
}
