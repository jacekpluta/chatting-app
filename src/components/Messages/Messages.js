import MessagesForm from "./MessagesForm";
import MessagesHeader from "./MessagesHeader";
import TypingLoader from "./TypingLoader";

import { Segment, Comment } from "semantic-ui-react";
import firebase from "../Firebase";
import React, { useState, useEffect, useRef } from "react";

import { connect } from "react-redux";
import { setUserPosts, setCurrentChannel } from "../../actions";
import SkeletonMessages from "./SkeletonMessages";

import Message from "./Message";
const Messages = props => {
  const [channelsRef] = useState(firebase.database().ref("channels"));
  const [privateMessagesRef] = useState(
    firebase.database().ref("privateMessages")
  );
  const [messagesRef] = useState(firebase.database().ref("messages"));
  const [usersRef] = useState(firebase.database().ref("users"));

  const messagesEndRef = useRef(null);

  const [friendsList, setFriendsList] = useState([]);
  const [messageImageLoading, setMessageImageLoading] = useState(false);
  const [allChannelMessages, setAllChannelMessages] = useState([]);
  const [messagesLoaded, setMessagesLoaded] = useState(false);
  const [searchTerm, setSearchTerm] = useState("");
  const [searchLoading, setSearchLoading] = useState(false);
  const [searchResult, setSearchResult] = useState([]);
  const [channelStarred, setChannelStarred] = useState(false);
  const [noMessages, setNoMessages] = useState(false);
  const [friendAdded, setFriendAdded] = useState(false);

  const {
    currentChannel,
    currentUser,
    isPrivateChannel,
    userTyping,
    usersList
  } = props;

  const setMessageImageLoadingTrue = () => {
    setMessageImageLoading(true);
  };

  const setMessageImageLoadingFalse = () => {
    setMessageImageLoading(false);
  };

  //RUNS AND REMOVES LISTENERS
  useEffect(() => {
    if (currentChannel && currentUser) {
      addListeners();

      return () => {
        const ref = getMessagesRef();
        ref.child(currentChannel.id).off("child_added");
        messagesRef.off();
      };
    }
  }, []);

  //COUNTS ALL USER POSTS IN CURRENT CHANNEL
  useEffect(() => {
    if (messagesLoaded) {
      countUserPosts();
    }
  }, [messagesLoaded]);

  const countUserPosts = () => {
    if (allChannelMessages.loadedMessages) {
      let userPosts = allChannelMessages.loadedMessages.reduce(
        (acc, message) => {
          if (message.currentUser.name in acc) {
            acc[message.currentUser.name].count += 1;
          } else {
            acc[message.currentUser.name] = {
              avatar: message.currentUser.avatar,
              count: 1
            };
          }
          return acc;
        },
        {}
      );

      props.setUserPosts(userPosts);
    } else {
      props.setUserPosts(null);
    }
  };

  //LISTENERS
  const addListeners = () => {
    addMessageListeners();
    loadData();
    addUserStarsListeners(currentChannel.id, currentUser.uid);
    addUserFriendsListeners(currentChannel.id, currentUser.uid);
  };

  //UPDATES CURRENT CHANNEL INFO
  useEffect(() => {
    if (messagesLoaded && !isPrivateChannel) {
      if (currentChannel.createdBy.uid === currentUser.uid) {
        const newCurrentChannelInfo = {
          createdBy: {
            avatar: currentUser.photoURL,
            name: currentChannel.createdBy.name,
            uid: currentChannel.createdBy.uid
          },
          details: currentChannel.details,
          id: currentChannel.id,
          name: currentChannel.name
        };

        channelsRef.child(currentChannel.id).update(newCurrentChannelInfo);
        props.setCurrentChannel(newCurrentChannelInfo);
      }
    }
  }, [messagesLoaded]);

  //ADD STARRED CHANNEL
  useEffect(() => {
    if (channelStarred) {
      usersRef.child(`${currentUser.uid}/starred`).update({
        [currentChannel.id]: {
          name: currentChannel.name,
          details: currentChannel.details,
          createdBy: {
            uid: currentChannel.createdBy.uid,
            name: currentChannel.createdBy.name,
            avatar: currentChannel.createdBy.avatar
          }
        }
      });
    }
  }, [channelStarred]);

  const handleStarred = () => {
    setChannelStarred(true);
  };

  //REMOVE STARRED CHANNEL
  const unstarChannel = () => {
    usersRef
      .child(`${currentUser.uid}/starred`)
      .child(currentChannel.id)
      .remove(err => {
        if (err !== null) {
          console.log(err);
        }
      });
    setChannelStarred(false);
  };

  //LISTEN FOR STARRED CHANNELS
  const addUserStarsListeners = (channelId, userId) => {
    usersRef
      .child(userId)
      .child("starred")
      .once("value")
      .then(data => {
        if (data.val() !== null) {
          const channelIds = Object.keys(data.val());
          const prevStarred = channelIds.includes(channelId);
          setChannelStarred(prevStarred);
        }
      });
  };

  //ADD FRIEND
  useEffect(() => {
    if (friendAdded) {
      let selectPrivateChannel = currentChannel.id.replace(currentUser.uid, "");
      let privateChannel = selectPrivateChannel.replace("/", "");

      usersRef.child(`${currentUser.uid}/friends`).update({
        [privateChannel]: {
          channelId: currentChannel.id,
          name: currentChannel.name,
          status: currentChannel.status,
          photoURL: currentChannel.photoURL
        }
      });
    }
  }, [friendAdded]);

  const handleAddFriend = () => {
    setFriendAdded(true);
  };

  //REMOVE FRIEND
  const unfriendPerson = () => {
    let mystring = currentChannel.id.replace(currentUser.uid, "");
    let privateChannel = mystring.replace("/", "");
    usersRef
      .child(`${currentUser.uid}/friends`)
      .child(privateChannel)
      .remove(err => {
        if (err !== null) {
          console.log(err);
        }
      });
    setFriendAdded(false);
  };

  //LISTEN FOR FRIRENDS
  const addUserFriendsListeners = (channelId, userId) => {
    let mystring = currentChannel.id.replace(currentUser.uid, "");
    let privateChannel = mystring.replace("/", "");

    usersRef
      .child(userId)
      .child("friends")
      .once("value")
      .then(data => {
        if (data.val() !== null) {
          const friendIds = Object.keys(data.val());
          const prevAdded = friendIds.includes(privateChannel);
          setFriendAdded(prevAdded);
          setFriendsList(allFriendsIds => [...allFriendsIds, friendIds]);
        }
      });
  };

  useEffect(() => {
    updateFriendsStatusListeners();
  }, [usersList]);

  const updateFriendsStatusListeners = () => {
    if (usersList && friendsList[0]) {
      const numberOfFriends = friendsList.length - 1;
      for (let i = 0; i < friendsList.length; i++) {
        usersList.forEach(element => {
          if (friendsList[numberOfFriends].includes(element.uid)) {
            let selectPrivateChannel = element.uid.replace(currentUser.uid, "");
            let privateChannel = selectPrivateChannel.replace("/", "");

            usersRef.child(`${currentUser.uid}/friends`).update({
              [privateChannel]: {
                channelId: element.uid,
                name: element.name,
                status: element.status,
                photoURL: element.photoURL
              }
            });
          }
        });
      }
    }
  };
  //SEARCH BAR
  const handleSearchChange = event => {
    setSearchTerm(event.target.value);
    setSearchLoading(true);
  };

  useEffect(() => {
    if (searchTerm) {
      handleSearchMessages();
    }
  }, [searchTerm]);

  const handleSearchMessages = () => {
    const channelMessages = [...allChannelMessages.loadedMessages];
    const regex = new RegExp(searchTerm, "gi");
    const searchResults = channelMessages.reduce((acc, message) => {
      if (
        (message.content && message.content.match(regex)) ||
        message.currentUser.name.match(regex)
      ) {
        acc.push(message);
      }
      return acc;
    }, []);

    setSearchResult(searchResults);
    setTimeout(() => {
      setSearchLoading(false);
    }, 1000);
  };

  //GET CURRENT MESSAGE REF
  const getMessagesRef = () => {
    return isPrivateChannel ? privateMessagesRef : messagesRef;
  };

  //DISPLAYS CURRENT CHANNEL NAME
  const displayChannelName = () => {
    if (currentChannel && !isPrivateChannel) {
      return `#${currentChannel.name}`;
    } else if (currentChannel && isPrivateChannel) {
      return `@${currentChannel.name}`;
    } else {
      return "";
    }
  };

  //LISTEN FOR MESSAGES
  const addMessageListeners = () => {
    let loadedMessages = [];
    const ref = getMessagesRef();

    ref.child(currentChannel.id).on("child_added", snapshot => {
      loadedMessages.push(snapshot.val());

      setAllChannelMessages({ loadedMessages });
      setNoMessages(false);
    });
  };

  ////CHECKS IF MESSAGES WERE LOADED IN CURRENT CHANNEL
  const wereMessagesLoaded = () => {
    if (!messagesLoaded) {
      setNoMessages(true);
    } else {
      setNoMessages(false);
    }
  };

  //LOADS MESSAGES ONCE ON CHANNEL START
  const loadData = () => {
    messagesRef
      .child(currentChannel.id)
      .limitToLast(200)
      .once("value", snapshot => {})
      .then(snapshot => {
        if (snapshot) {
          if (snapshot.val()) {
          } else {
            wereMessagesLoaded();
          }
        }
      })
      .then(() => {
        setMessagesLoaded(true);
      });
  };

  //RENDERS ALL CHANNEL MESSAGES
  const renderMessages = () =>
    allChannelMessages.loadedMessages.map(message => {
      return (
        <Message
          key={message.timeStamp}
          message={message}
          currentUser={currentUser}
          messageImageLoading={messageImageLoading}
          searchResult={searchResult}
          searchTerm={searchTerm}
        />
      );
    });

  //SCROLL TO THE BOTTOM OF CHANNEL WINDOW
  const scrollToBottom = () => {
    messagesEndRef.current.scrollIntoView({ behavior: "auto" }); //smooth
  };

  useEffect(() => {
    if (messagesLoaded === true) {
      const timer = setTimeout(() => {
        scrollToBottom();
      }, 500);
      return () => clearTimeout(timer);
    }
  }, [messagesLoaded]);

  useEffect(() => {
    if (userTyping) scrollToBottom();
    else scrollToBottom();
  }, [userTyping]);
  ////////////////////////////////////////////////////////////

  return (
    <React.Fragment>
      <MessagesHeader
        handleSearchChange={handleSearchChange}
        displayChannelName={displayChannelName}
        searchLoading={searchLoading}
        isPrivateChannel={isPrivateChannel}
        handleStarred={handleStarred}
        unstarChannel={unstarChannel}
        channelStarred={channelStarred}
        friendAdded={friendAdded}
        handleAddFriend={handleAddFriend}
        unfriendPerson={unfriendPerson}
      ></MessagesHeader>

      <Segment>
        <Comment.Group className="messages">
          {/*shows messages skeleton on messages loading then checks if there is a messages in channel and displays "no messages" or shows all messages*/}
          {allChannelMessages.loadedMessages &&
          searchTerm === "" &&
          messagesLoaded
            ? renderMessages()
            : [
                noMessages
                  ? "No messages in the channel"
                  : [
                      allChannelMessages.loadedMessages &&
                      searchTerm === "" &&
                      messagesLoaded
                        ? renderMessages()
                        : [searchTerm === "" ? <SkeletonMessages /> : ""]
                    ]
              ]}

          {/*displays user name + usertypeing animation*/}

          {userTyping && userTyping.isUserTyping
            ? userTyping.userTypingUid !== currentUser.uid &&
              "User " + userTyping.userTypingName + " is writing"
            : ""}

          {userTyping && userTyping.isUserTyping
            ? userTyping.userTypingUid !== currentUser.uid && (
                <TypingLoader userTyping={userTyping}></TypingLoader>
              )
            : ""}

          {allChannelMessages.loadedMessages && searchTerm
            ? searchResult.map(message => {
                return (
                  <Message
                    key={message.timeStamp}
                    message={message}
                    currentUser={currentUser}
                    messageImageLoading={messageImageLoading}
                    searchResult={searchResult}
                    searchTerm={searchTerm}
                  />
                );
              })
            : ""}
          <div ref={messagesEndRef} />
        </Comment.Group>
      </Segment>
      <MessagesForm
        setMessageImageLoadingTrue={setMessageImageLoadingTrue}
        setMessageImageLoadingFalse={setMessageImageLoadingFalse}
        messagesRef={messagesRef}
        currentChannel={currentChannel}
        currentUser={currentUser}
        messageImageLoading={messageImageLoading}
        getMessagesRef={getMessagesRef}
      ></MessagesForm>
    </React.Fragment>
  );
};

export default connect(null, { setUserPosts, setCurrentChannel })(Messages);
