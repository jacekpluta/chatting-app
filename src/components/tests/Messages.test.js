import React from "react";

import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Segment, CommentGroup } from "semantic-ui-react";
import SkeletonMessages from "../Messages/SkeletonMessages";
import Messages from "../../components/Messages/Messages";
import Redux from "../../Redux";
import { Steps } from "intro.js-react";
import renderer from "react-test-renderer";
import MessagesHeader from "../Messages/MessagesHeader";
import MessagesForm from "../Messages/MessagesForm";

Enzyme.configure({
  adapter: new Adapter(),
});

let wrapper;

beforeEach(() => {
  //   let scrollIntoViewMock = jest.fn();
  //   window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

  wrapper = shallow(
    <Redux>
      <Messages />
    </Redux>
  )
    .dive()
    .dive()
    .dive();
});

it("should find one Segment component", () => {
  expect(wrapper.find(Segment).length).toEqual(1);
});

it("should find one Steps component", () => {
  expect(wrapper.find(Steps).length).toEqual(1);
  expect(wrapper.find(Steps).prop("enabled")).toEqual(false);
  expect(wrapper.find(Steps).prop("initialStep")).toEqual(0);
  expect(wrapper.find(Steps).prop("onExit")).toEqual(expect.any(Function));
});
it("should find one CommentGroup component with class name `messages`", () => {
  expect(wrapper.find(CommentGroup).length).toEqual(1);
  expect(wrapper.find(CommentGroup).hasClass("messages")).toEqual(true);
});

it("should find one MessagesForm component", () => {
  expect(wrapper.find(MessagesForm).length).toEqual(1);
});

it("should find one SkeletonMessages component", () => {
  expect(wrapper.find(SkeletonMessages).length).toEqual(1);
});

it("should find one MessagesHeader component", () => {
  expect(wrapper.find(MessagesHeader).length).toEqual(1);
});

it("should Steps component when Steps prop is enabled", () => {
  const wrapper = shallow(
    <Redux>
      <Messages stepsEnabled={true} />
    </Redux>
  )
    .dive()
    .dive()
    .dive();

  expect(wrapper.find(Steps).prop("enabled")).toEqual(true);
});

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Redux>
        <Messages />
      </Redux>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
