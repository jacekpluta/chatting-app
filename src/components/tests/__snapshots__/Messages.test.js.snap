// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="ui clearing segment messagesHeader"
  >
    <h3
      className="ui left floated header"
      fluid="true"
      style={
        Object {
          "paddingTop": "16px",
        }
      }
    >
      <span
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
         
        
         
        <i
          aria-hidden="true"
          className="star outline icon iconColor"
          onClick={[Function]}
        />
      </span>
    </h3>
    <h3
      className="ui right floated center aligned header"
      fluid="true"
      style={
        Object {
          "paddingTop": "6px",
        }
      }
    >
      <span>
        <i
          aria-hidden="true"
          className="question circle icon iconColor"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        />
          
            
        <div
          className="ui icon input"
          style={
            Object {
              "width": "170px",
            }
          }
        >
          <input
            name="searchTerm"
            onChange={[Function]}
            placeholder="search "
            type="text"
          />
          <i
            aria-hidden="true"
            className="search icon"
          />
        </div>
      </span>
    </h3>
  </div>,
  <div
    className="ui segment"
  >
    <div
      className="ui comments messages"
    >
      
      <div>
        <div
          className="ui two column grid"
        >
          <div
            className="row"
          >
            <div
              className="two wide column"
            >
              <span>
                <span
                  className="css-19bon8n react-loading-skeleton"
                  style={
                    Object {
                      "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              className="three wide column"
              style={
                Object {
                  "paddingLeft": -5,
                }
              }
            >
              <div
                className="row"
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                className="row"
                style={
                  Object {
                    "paddingTop": 5,
                  }
                }
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                        "height": 20,
                        "width": 250,
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="ui two column grid"
        >
          <div
            className="row"
          >
            <div
              className="two wide column"
            >
              <span>
                <span
                  className="css-19bon8n react-loading-skeleton"
                  style={
                    Object {
                      "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              className="three wide column"
              style={
                Object {
                  "paddingLeft": -5,
                }
              }
            >
              <div
                className="row"
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                className="row"
                style={
                  Object {
                    "paddingTop": 5,
                  }
                }
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                        "height": 20,
                        "width": 250,
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="ui two column grid"
        >
          <div
            className="row"
          >
            <div
              className="two wide column"
            >
              <span>
                <span
                  className="css-19bon8n react-loading-skeleton"
                  style={
                    Object {
                      "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              className="three wide column"
              style={
                Object {
                  "paddingLeft": -5,
                }
              }
            >
              <div
                className="row"
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                className="row"
                style={
                  Object {
                    "paddingTop": 5,
                  }
                }
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                        "height": 20,
                        "width": 250,
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="ui two column grid"
        >
          <div
            className="row"
          >
            <div
              className="two wide column"
            >
              <span>
                <span
                  className="css-19bon8n react-loading-skeleton"
                  style={
                    Object {
                      "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      "height": 40,
                      "width": 40,
                    }
                  }
                >
                  ‌
                </span>
              </span>
            </div>
            <div
              className="three wide column"
              style={
                Object {
                  "paddingLeft": -5,
                }
              }
            >
              <div
                className="row"
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
              <div
                className="row"
                style={
                  Object {
                    "paddingTop": 5,
                  }
                }
              >
                <span>
                  <span
                    className="css-19bon8n react-loading-skeleton"
                    style={
                      Object {
                        "animation": "animation-16jpnkj 1.2s ease-in-out infinite",
                        "height": 20,
                        "width": 250,
                      }
                    }
                  >
                    ‌
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      
       
      
      
      
      
      <div />
    </div>
  </div>,
  <div
    className="ui segment messageForm"
  >
    
     
    <div
      className="ui fluid input"
      style={
        Object {
          "marginBottom": "0.7 em",
        }
      }
    >
      <button
        className="ui grey small icon button"
        onClick={[Function]}
        role="button"
      >
        <i
          aria-hidden="true"
          className="upload icon"
        />
      </button>
      <button
        className="ui grey small icon button"
        onClick={[Function]}
        role="button"
      >
        <i
          aria-hidden="true"
          className="smile outline icon"
        />
      </button>
      <input
        autoFocus={true}
        name="message"
        onChange={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="Write your message"
        type="text"
        value=""
      />
      <button
        className="ui grey small icon button"
        onClick={[Function]}
        role="button"
      >
        <i
          aria-hidden="true"
          className="arrow alternate circle right outline icon"
        />
      </button>
    </div>
    
  </div>,
]
`;
